module/builder/collection/builder.go

```diff
@@ -222,32 +218,38 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 
 			// ensure the reference block is not too old
 			txID := tx.ID()
-			if refChainFinalizedHeight-refHeader.Height > uint64(flow.DefaultTransactionExpiry-b.expiryBuffer) {
+			if refChainFinalizedHeight-refHeader.Height > uint64(flow.DefaultTransactionExpiry-b.config.ExpiryBuffer) {
 				// the transaction is expired, it will never be valid
 				b.transactions.Rem(txID)
 				continue
 			}
 
 			// check that the transaction was not already used in un-finalized history
-			_, duplicated := unfinalizedLookup[txID]
-			if duplicated {
+			if lookup.isUnfinalizedAncestor(txID) {
 				continue
 			}
 
 			// check that the transaction was not already included in finalized history.
-			_, duplicated = finalizedLookup[txID]
-			if duplicated {
+			if lookup.isFinalizedAncestor(txID) {
 				// remove from mempool, conflicts with finalized block will never be valid
 				b.transactions.Rem(txID)
 				continue
 			}
 
+			// enforce rate limiting rules
+			if limiter.shouldRateLimit(tx) {
```

Comment:
I am guess rate limiting a tx would be an exception vs a norm and if that is the case then should we log.debug the tx and payer we rate limited to figure out who is abusing the system from logs?

Reply:
Alternatively, we would add a metric later "rate_limited_tx".

Reply:
Good point. I won't have time before the spork we're trying to get this in for, but I'll add a note to the other issue.